lib/client/conntest/connection_tester.go

```diff
@@ -22,16 +22,48 @@ import (
 
 	"github.com/gravitational/trace"
 
+	"github.com/gravitational/teleport/api/client/proto"
 	"github.com/gravitational/teleport/api/defaults"
 	"github.com/gravitational/teleport/api/types"
 	"github.com/gravitational/teleport/lib/auth"
+	wanlib "github.com/gravitational/teleport/lib/auth/webauthn"
 )
 
+type MfaResponse struct {
+	// TotpCode is a code for a otp device.
+	TotpCode string `json:"totpCode"`
+	// WebauthnResponse is a response from a webauthn device.
+	WebauthnResponse *wanlib.CredentialAssertionResponse `json:"webauthnResponse"`
+}
+
+func (r *MfaResponse) getMFAResponseProtoReq() (*proto.MFAAuthenticateResponse, error) {
+	if r.TotpCode != "" && r.WebauthnResponse != nil {
+		return nil, trace.BadParameter("only one MFA response field can be set")
+	}
+
+	if r.TotpCode != "" {
+		return &proto.MFAAuthenticateResponse{Response: &proto.MFAAuthenticateResponse_TOTP{
+			TOTP: &proto.TOTPResponse{Code: r.TotpCode},
+		}}, nil
+	}
+
+	if r.WebauthnResponse != nil {
+		return &proto.MFAAuthenticateResponse{Response: &proto.MFAAuthenticateResponse_Webauthn{
+			Webauthn: wanlib.CredentialAssertionResponseToProto(r.WebauthnResponse),
+		}}, nil
+	}
+
+	// Can be nil since it's optional.
+	return nil, nil
+}
+
 // TestConnectionRequest contains
 // - the identification of the resource kind and resource name to test
 // - additional paramenters which depend on the actual kind of resource to test
 // As an example, for SSH Node it also includes the User/Principal that will be used to login.
 type TestConnectionRequest struct {
+	// MFAResponse is an optional field that holds a response to a MFA device challenge.
+	MFAResponse MfaResponse `json:"mfa_response"`
```

Comment:
I noticed that MfaResponse uses "lowerCamel" for the JSON tag, but here we use "lower_snake". Is there a standard for this?

Reply:
I wish there was for at least the web layer because JS convention is camel casing, so snake casing requires more tedious processing in the front (when front gets snake requests, we have to convert to camel. Camel case gets converted to snake responses). 

History for the web layer is that we started with snake case, then we (alexey and I) started converting to camel casing, and then more people joined and snake casing came back. Which makes sense because we use snake casing everywhere else in the back and it's hard to enforce. 

I'll stick to snaking from now on since it seems to be the easier way to keep things consistent 🥲 I don't think this will matter in the future since there were talks of someday (🌈 ) to transition into grpc

Reply:
Yeah, I know I'm guilty of some snake_casing just because I followed what was there, or didn't consider it enough.

> someday (🌈 ) to transition into grpc

The dream. 🤞